<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">

  <title><![CDATA[Category: tech | xixitalk's snippet]]></title>
  <link href="http://xixitalk.github.io/blog/categories/tech/atom.xml" rel="self"/>
  <link href="http://xixitalk.github.io/"/>
  <updated>2016-09-30T02:06:00+00:00</updated>
  <id>http://xixitalk.github.io/</id>
  <author>
    <name><![CDATA[xixitalk]]></name>
    
  </author>
  <generator uri="http://octopress.org/">Octopress</generator>

  
  <entry>
    <title type="html"><![CDATA[我的AStyle配置选项]]></title>
    <link href="http://xixitalk.github.io/blog/2016/07/11/my-AStyle-config/"/>
    <updated>2016-07-11T18:35:57+00:00</updated>
    <id>http://xixitalk.github.io/blog/2016/07/11/my-AStyle-config</id>
    <content type="html"><![CDATA[<p>我的AStyle代码格式工具的选项：AStyle.exe -A1 -C -S -K -Y -f -s4 -p -U -o -n main.c</p>

<p>在<code>notepad++</code>里添加：<code>运行(R)--运行(R)</code> 选择AStyle.exe，选项输入<code>-A1 -C -S -K -Y -f -s4 -p -U -o -n "$(FULL_CURRENT_PATH)"</code>，然后点击<code>保存</code>，这样就保存在运行菜单里了。每次使用从<code>运行(R)</code>点击即可。</p>

<!--more-->

<p>简略选项：<code>-A1 -C -S -K -Y -f -s4 -p -U -o -n</code>
对应长选项如下：</p>

<p><code>
--style=bsd
--indent-classes
--indent-switches
--indent-cases
--indent-col1-comments
--break-blocks
--indent=spaces=4
--pad-oper
--unpad-paren
--keep-one-line-statements
--suffix=none
</code></p>

]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[UINT32引起的死循环]]></title>
    <link href="http://xixitalk.github.io/blog/2016/07/09/dead-loop-UINT32/"/>
    <updated>2016-07-09T10:02:35+00:00</updated>
    <id>http://xixitalk.github.io/blog/2016/07/09/dead-loop-UINT32</id>
    <content type="html"><![CDATA[<p>UINT32 引起的死循环</p>

<!--more-->

<p>上周出差定位一个死循环。一个函数进行一个信号量的初始化，每次进行一次业务都要调用这个函数初始化，这个函数作用是保障初始化后信号量初始化值为<code>SEM_INIT_NUM</code>，但因为<code>UINT32</code>的关系出现死循环了。代码如下：</p>

<script src="https://gist.github.com/xixitalk/5c176a78a8743465061ef15ab5f9a248.js"></script>

<p>出现问题的时候，<code>semCount</code>为4，而宏<code>SEM_INIT_NUM</code>为3，造成<code>Count = SEM_INIT_NUM - semCount</code>为<code>-1</code>了，但因为<code>Count</code>是<code>UINT32</code>，从而变成<code>0xFFFFFFFF</code>，一个极大值，造成for循环执行长时间不退出。修改方法是将<code>Count</code>的<code>UINT32</code>改成<code>SINT32</code>。这是一个非常低级的代码错误。</p>

<p>其实更简单的是删除信号量，重新创建信号量，这样代码会容易读的多。</p>

<p><code>
DeleteSemaphore(ptxSem);
ptxSem = CreateSemaphore(SEM_INIT_NUM);
</code></p>
]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[免费申请S/MIME证书]]></title>
    <link href="http://xixitalk.github.io/blog/2014/06/21/Free-SMIME-Certificate-Key/"/>
    <updated>2014-06-21T00:05:11+00:00</updated>
    <id>http://xixitalk.github.io/blog/2014/06/21/Free-SMIME-Certificate-Key</id>
    <content type="html"><![CDATA[<h2 id="section">申请证书</h2>
<p>从这里可以申请到免费的S/MIME 邮件证书：<a href="http://www.instantssl.com/ssl-certificate-products/free-email-certificate.html">http://www.instantssl.com/ssl-certificate-products/free-email-certificate.html</a>，
提交生成之后会把证书发送到邮箱里。</p>

<h2 id="section-1">证书安装</h2>
<p>在邮箱里点击连接就会把证书安装到电脑系统中（包括windows和Mac OSX)。<br />
证书认证关系如下：  </p>

<ul>
  <li>第一级根证书USERTrust(UTN-USERFirst-Client Authentication and Email)，系统中默认都带了；</li>
  <li>第二级中级证书是COMODO Client Authentication and Secure Email CA;</li>
  <li>第三级才是刚刚申请的个人证书。<br />
如果在证书管理器查看刚刚申请的个人证书显示不正常的话，说明系统中缺少第二级中级证书，从这里下载安装<a href="http://crt.comodoca.com/COMODOClientAuthenticationandSecureEmailCA.crt">http://crt.comodoca.com/COMODOClientAuthenticationandSecureEmailCA.crt</a>。</li>
</ul>

<h2 id="ios">证书安装到iOS系统</h2>
<p>在电脑系统里打开证书管理器（windows系统是运行certmgr.msc），把证书（包含私钥）导出，windows是pfx格式，Mac OSX是p12格式，导出的证书要设置私钥保护密码，iOS的证书安装需要输入这个密码。把这个pfx或者p12格式的证书发邮件到iOS上，点击打开即提示安装，输入之前的保护密码就安装到iOS系统中了。<br />
同样，iOS系统中缺少第二级中级证书，从这里下载安装<a href="http://crt.comodoca.com/COMODOClientAuthenticationandSecureEmailCA.crt">http://crt.comodoca.com/COMODOClientAuthenticationandSecureEmailCA.crt</a>。</p>

<!--more-->

<h2 id="ios-1">iOS里启用加密和签名</h2>
<p>在iOS-邮件、通讯录、日历-邮件帐户-高级 启用S/MIME，选择上面安装的证书。iOS自带的mail程序发邮件即可自动进行加密和签名。</p>

<h2 id="section-2">群发公钥证书</h2>
<p>在电脑证书管理里导出证书（不包含私钥），windows是cer文件，把这个文件群发出来告诉朋友。别人给你发邮件就用这个证书进行加密。</p>

<h2 id="smime">我的S/MIME证书</h2>
<p>我的S/MIME证书<a href="http://xixitalk.github.io/static/pub.cer">http://xixitalk.github.io/static/pub.cer</a>,欢迎用S/MIME给我发邮件。你的系统也可能需要安装第二级COMODO的证书，从这里下载安装<a href="http://crt.comodoca.com/COMODOClientAuthenticationandSecureEmailCA.crt">http://crt.comodoca.com/COMODOClientAuthenticationandSecureEmailCA.crt</a>。</p>

<h2 id="section-3">其他参考</h2>
<p>Mac OSX导出证书到iOS参考<a href="http://feinstruktur.com/blog/2011/12/12/using-smime-on-ios-devices.html">http://feinstruktur.com/blog/2011/12/12/using-smime-on-ios-devices.html</a></p>
]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[curl下载断点传续]]></title>
    <link href="http://xixitalk.github.io/blog/2014/04/12/curl-continue/"/>
    <updated>2014-04-12T17:22:17+00:00</updated>
    <id>http://xixitalk.github.io/blog/2014/04/12/curl-continue</id>
    <content type="html"><![CDATA[<h3 id="curl">curl下载断点传续</h3>

<p><code>
curl -C offset -O http://google.com/p4.mp4 
</code></p>

<p>offset手动输入已经下载的字节数</p>

<p><code>
curl -C - -O http://google.com/p4.mp4 
</code></p>

<p><code>-C -</code>可以自动识别offet续传</p>

<!--more-->

<h3 id="curl-1">curl使用代理</h3>

<p><code>
curl -x 192.168.1.106:8118  http://google.com/p4.mp4 
</code></p>

<p><code>-x</code> 后面跟HTTP代理的host和port</p>

<p><code>
curl  --socks5-hostname 192.168.1.106:9050  http://google.com/p4.mp4
</code></p>

<p><code>--socks5-hostname</code>后面跟socks5代理的host和port</p>

]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[twitter-python BadStatusLine错误]]></title>
    <link href="http://xixitalk.github.io/blog/2014/04/06/twitter-python-badstatusline/"/>
    <updated>2014-04-06T20:20:25+00:00</updated>
    <id>http://xixitalk.github.io/blog/2014/04/06/twitter-python-badstatusline</id>
    <content type="html"><![CDATA[<p>事情是这样的：我用<a href="https://github.com/bear/python-twitter">twitter-python</a>在本地运行， 抓取一个推号的homeline， 抓取后sleep两分钟再抓取， 不停的循环｡从昨天早上开始， 程序出现如下错误， 且每次都是这个错误｡获得推友收藏列表和fo人的api返回都是正常的，就是REST的api出这种错误。</p>

<!--more-->

<p><code>
Traceback (most recent call last):
  File "hottweetdo.py", line 183, in &lt;module&gt;
    main()
  File "hottweetdo.py", line 126, in main
    status = api2.GetHomeTimeline(since_id=since_id1,count=200)
  File "/media/westdata/read/python-twitter-master/twitter/api.py", line 541, in GetHomeTimeline
    json = self._RequestUrl(url, 'GET', data=parameters)
  File "/media/westdata/read/python-twitter-master/twitter/api.py", line 3383, in _RequestUrl
    timeout=self._requests_timeout
  File "/usr/local/lib/python2.7/dist-packages/requests/api.py", line 55, in get
    return request('get', url, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/api.py", line 44, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 383, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 486, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 382, in send
    raise ConnectionError(e)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/home_timeline.json?count=200 (Caused by &lt;class 'httplib.BadStatusLine'&gt;: '')
</code></p>

<p>代码很简单，就是用四个key生成一个api对象，然后直接调用<code>GetHomeTimeline</code></p>

<p>我google之后， 看到很多类似的分析， 但是我都尝试了， 还是没有解决｡尝试罗列如下｡</p>

<h3 id="bad-proxy-server">bad proxy server</h3>
<p><a href="https://stackoverflow.com/questions/18478013/python-requests-exceptions-connectionerror-max-retries-exceeded-witth-url">有人说是bad proxy server</a>,程序确实使用代理了，代理是shadowsocks+polipo， 代理设置在环境变量HTTP_PROXY和HTTPS_PROXY｡我尝试了不同地区的代理， 甚至切换了不同卖家的代理， 依然出错｡</p>

<h3 id="dns">DNS问题</h3>
<p>尝试了不同的DNS server不起作用</p>

<h3 id="httplib2-08-077">httplib2 0.8的问题， 使用0.7.7</h3>
<p>不起作用</p>

<h3 id="twitter-api">twitter API的问题</h3>
<p>twitter-python的一个<a href="https://github.com/bear/python-twitter/issues/131">issues解答</a>说是twitter API诡异的问题｡问题是我只能等待吗？</p>

<h3 id="maxretries5">设置maxretries为5</h3>
<p><a href="https://stackoverflow.com/questions/15431044/can-i-set-max-retries-for-requests-request/15431343#15431343">https://stackoverflow.com/questions/15431044/can-i-set-max-retries-for-requests-request/15431343#15431343</a> 不起作用</p>

<h3 id="requestssession-keep-alivefalse">设置requests的session keep-alive为False</h3>
<p>不起作用</p>

<p>请python高手或者twitter api高手指点迷津，可以直接在下面留言，或者推上@我。</p>

<h2 id="section">[更新后记]</h2>

<p>目前解决的办法：1、重新新建一个API，用新的consumer_key和access token；2、更换代理，换用tor+polipo，tor的IP变化比较频繁。经过上面操作，错误暂时不再出现了。</p>

<p>问题可能原因：应该是twitter服务器认为API滥用了，当get提交给服务器后，twitter服务器只返回了ACK，接下来没有返回任何数据，所以httplib判断为BadStatusLine，因为没有数据，所以连status号都没有。</p>

<h2 id="section-1">[再次更新后记]</h2>
<p>每天生成HTMLpush到github后自动发推又出现BadStatusLine了，所幸不是<code>GetHomeTimeline</code>。</p>

<p><code>
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/update.json (Caused by &lt;class 'httplib.BadStatusLine'&gt;: '')
</code></p>
]]></content>
  </entry>
  
</feed>
